File: Remote/Source/Bank/Bankables/ImageView.m
//
//  ImageView.m
//  Remote
//
//  Created by Jason Cardwell on 8/24/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

#import "ImageView.h"
#import "Remote-Swift.h"


@interface ImageView ()

@property (nonatomic, strong) UIImage * colorImage;
@property (nonatomic, strong) UIImage * rawImage;

@end

@implementation ImageView

@dynamic color, image;
@synthesize colorImage = _colorImage, rawImage = _rawImage;

+ (instancetype)imageViewWithImage:(Image *)image color:(UIColor *)color {
  if (!image) ThrowInvalidNilArgument(image);
  ImageView * imageView = [self createInContext:image.managedObjectContext];
  imageView.image = image;
  imageView.color = color;
  return imageView;
}

- (UIImage *)colorImage {
  if (self.rawImage && self.color && !_colorImage)
    self.colorImage = [self.rawImage recoloredImageWithColor:self.color];
  return _colorImage;
}

- (UIImage *)rawImage {
  if (self.image && !_rawImage) self.rawImage = self.image.image;
  return _rawImage;
}

- (void)setImage:(Image *)image {
  [self willChangeValueForKey:@"image"];
  [self setPrimitiveValue:image forKey:@"image"];
  [self didChangeValueForKey:@"image"];
  self.rawImage   = nil;
  self.colorImage = nil;
}

- (void)setColor:(UIColor *)color {
  [self willChangeValueForKey:@"color"];
  [self setPrimitiveValue:color forKey:@"color"];
  [self didChangeValueForKey:@"color"];
  self.colorImage = nil;
}

@end

File: Remote/Source/Remote Elements/Control State Sets/ImageView.swift
//
//  ImageView.swift
//  Remote
//
//  Created by Jason Cardwell on 10/3/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

import Foundation
import CoreData
import MoonKit

@objc(ImageView)
class ImageView: ModelObject {

  @NSManaged var color: UIColor?
  @NSManaged var image: Image

  @NSManaged var buttonIcon: Button?
  @NSManaged var buttonImage: Button?
  @NSManaged var imageSetDisabled: ControlStateImageSet?
  @NSManaged var imageSetDisabledSelected: ControlStateImageSet?
  @NSManaged var imageSetHighlighted: ControlStateImageSet?
  @NSManaged var imageSetHighlightedDisabled: ControlStateImageSet?
  @NSManaged var imageSetHighlightedSelected: ControlStateImageSet?
  @NSManaged var imageSetNormal: ControlStateImageSet?
  @NSManaged var imageSetSelected: ControlStateImageSet?
  @NSManaged var imageSetSelectedHighlightedDisabled: ControlStateImageSet?


  var rawImage: UIImage? { return image.image }

  var colorImage: UIImage? {
    if let img = rawImage {
      if let imgColor = color { return UIImage(fromAlphaOfImage: img, color: imgColor) }
      else { return img }
    } else {
      return nil
    }
  }

  /**
  updateWithData:

  :param: data [NSObject:AnyObject]!
  */
  override func updateWithData(data: [NSObject:AnyObject]!) {
    super.updateWithData(data)

    if let imageData = data["image"] as? [NSObject:AnyObject], let moc = managedObjectContext,
      let image = Image.importObjectFromData(imageData, context: moc) {
        self.image = image
    }

    if let colorJSON = data["color"] as? String, let color = UIColor(JSONValue: colorJSON) {
      self.color = color
    }

  }

  /**
  JSONDictionary

  :returns: MSDictionary!
  */
  override func JSONDictionary() -> MSDictionary {
    let dictionary = super.JSONDictionary()

    dictionary["image"] = image.commentedUUID
    if let color = self.color { dictionary["color"] = color.JSONValue }

    dictionary.compact()
    dictionary.compress()

    return dictionary
  }



}
